### PR TITLE
Validate Helm values.yaml

### PR DESCRIPTION
**Pull request by flvndh**
Tackle #10634

### CODE REVIEW

**Comment by mik-laj** [comment]
We have some errors Can you look at it?
```
2020-08-31T16:29:32.9918043Z helm install airflow . --namespace airflow --set defaultAirflowRepository=apache/airflow --set images.airflow.repository=apache/airflow --set images.airflow.tag=master-python3.6-kubernetes -v 1 --set defaultAirflowTag=master-python3.6-kubernetes -v 1 --set config.api.auth_backend=airflow.api.auth.backend.default
2020-08-31T16:29:33.1466658Z Error: values don't meet the specifications of the schema(s) in the following chart(s):
2020-08-31T16:29:33.1467262Z airflow:
2020-08-31T16:29:33.1467809Z - fernetKey: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1468655Z - statsd.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1469091Z - registry.secretName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1469800Z - rbacEnabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1470663Z - elasticsearch.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1471205Z - elasticsearch.secretName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1471649Z - cleanup.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1472085Z - allowPodLaunching: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1472753Z - postgresql.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1473407Z - data.metadataSecretName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1473915Z - data.resultBackendSecretName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1474567Z - data.metadataConnection.host: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1475291Z - data.resultBackendConnection.host: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1475828Z - pgbouncer.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1476508Z - pgbouncer.podDisruptionBudget.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1476875Z - fernetKeySecretName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1477513Z - ingress.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1478153Z - ingress.web.tls.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1478797Z - ingress.flower.tls.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1479444Z - workers.persistence.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1480094Z - workers.persistence.storageClassName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1480871Z - workers.persistence.fixPermissions: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1484488Z - workers.safeToEvict: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1492524Z - workers.keda.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1499109Z - scheduler.podDisruptionBudget.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1499838Z - scheduler.airflowLocalSettings: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1500356Z - scheduler.safeToEvict: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1501174Z - dags.persistence.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1501673Z - dags.persistence.storageClassName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1502131Z - dags.persistence.existingClaim: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1502553Z - dags.gitSync.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1502934Z - networkPolicies.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1503347Z - images.flower.tag: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1503757Z - images.flower.repository: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1504173Z - webserver.webserverConfig: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1504606Z - webserver.defaultUser.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1505087Z - redis.password: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1505510Z - redis.safeToEvict: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1505936Z - redis.persistence.enabled: Invalid type. Expected: integer, given: boolean
2020-08-31T16:29:33.1506365Z - redis.persistence.storageClassName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1506783Z - redis.passwordSecretName: Invalid type. Expected: string, given: null
2020-08-31T16:29:33.1507330Z - redis.brokerURLSecretName: Invalid type. Expected: string, given: null
```

File: chart/values.schema.json [hunk 1]
```diff
@@ -0,0 +1,1250 @@
...
+                        "containerName": {
+                            "description": "Git sync container name.",
+                            "type": "string"
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
```
Comment (mik-laj) [inline chart/values.schema.json]:
```suggestion
}

```

Comment (mik-laj) [reply chart/values.schema.json]: it's a fix for the static-check error.

File: chart/values.schema.json [hunk 2]
```diff
@@ -0,0 +1,1250 @@
...
+                            "type": "string"
+                        },
+                        "git_sync_credentials_secret": {
+                            "type": "string"
+                        }
+                    }
+                },
+                "kubernetes_secrets": {
+                    "type": "object",
+                    "properties": {
```
Comment (mik-laj) [inline chart/values.schema.json:1142]: It should be a map of string/string.

File: chart/values.schema.json [hunk 3]
```diff
@@ -0,0 +1,1250 @@
...
+                        "AIRFLOW__CORE__SQL_ALCHEMY_CONN": {
+                            "type": "string"
+                        },
+                        "AIRFLOW__CORE__FERNET_KEY": {
+                            "type": "string"
+                        }
+                    }
```
Comment (mik-laj) [inline chart/values.schema.json]: It should be a free-form object. We don't want to limit the name of the keys.
```suggestion
                    "additionalProperties": { "type": "string" }
```

File: chart/values.schema.json [hunk 4]
```diff
@@ -0,0 +1,1250 @@
...
+                "postgresqlUsername": {
+                    "description": "Postgresql username.",
+                    "type": "string"
+                }
+            }
+        },
+        "config": {
+            "description": "Settings to go into the mounted airflow.cfg",
+            "type": "object",
+            "properties": {
```
Comment (mik-laj) [inline chart/values.schema.json]: Here similar, but a little more complex.  We expect there will be a map, which has a map that contains arbitrary text strings. ``Map<String, Map<String, String>``

File: chart/values.schema.json [hunk 5]
```diff
@@ -0,0 +1,1250 @@
...
+            "type": "object"
+        },
+        "rbacEnabled": {
+            "description": "Enable RBAC (default on most clusters these days).",
+            "type": "boolean"
+        },
+        "executor": {
+            "description": "Airflow executor.",
+            "type": "string",
+            "enum": ["SequentialExecutor", "LocalExecutor", "CeleryExecutor", "KubernetesExecutor"]
```
Comment (mik-laj) [inline chart/values.schema.json]: Uses can provide any string to use custom executor.

> If it is to load your own executor, then you should specify the full path to the module e.g. my_acme_company.executors.MyCustomExecutor.

https://airflow.readthedocs.io/en/latest/executor/index.html

File: chart/values.schema.json [hunk 6]
```diff
@@ -0,0 +1,1250 @@
...
+            "description": "Select certain nodes for airflow pods.",
+            "type": "object"
+        },
+        "tolerations": {
+            "description": "Select certain nodes for airflow pods.",
+            "type": "array"
+        },
+        "labels": {
+            "description": "Add common labels to all objects and pods defined in this chart.",
+            "type": "object"
```
Comment (mik-laj) [inline chart/values.schema.json]:
```suggestion
            "type": "object",
            "additionalProperties": { "type": "string" }
```

File: chart/values.schema.json [hunk 7]
```diff
@@ -0,0 +1,1250 @@
...
+            "description": "Default airflow repository. Overrides all the specific images below.",
+            "type": "string"
+        },
+        "defaultAirflowTag": {
+            "description": "Default airflow tag to deploy.",
+            "type": "string"
+        },
+        "nodeSelector": {
+            "description": "Select certain nodes for airflow pods.",
+            "type": "object"
```
Comment (mik-laj) [inline chart/values.schema.json]:
```suggestion
            "type": "object",
            "additionalProperties": { "type": "string" }
```
? I'm not sure.

**Comment by mik-laj** [comment]
It would be fantastic if we could generate documentation for README.md based on this, but we should do it in a separate PR. It has a lot of potential.

File: chart/values.schema.json (hunk 7 above)
Comment (flvndh) [reply chart/values.schema.json]: You are right. The [doc](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) for `nodeSelector` says:
> `nodeSelector` is the simplest recommended form of node selection constraint. `nodeSelector` is a field of PodSpec. It specifies a map of key-value pairs.

File: chart/values.schema.json (hunk 3 above)
Comment (flvndh) [reply chart/values.schema.json]: Indeed. Change done.

**Comment by flvndh** [comment]
> It would be fantastic if we could generate documentation for README.md based on this, but we should do it in a separate PR. It has a lot of potential.

Indeed, this is a great way to keep both in sync. Let me know if I can help with this as well :)

File: chart/values.schema.json (hunk 4 above)
Comment (flvndh) [reply chart/values.schema.json]: When I do a
```sh
helm install airflow . --namespace airflow --set defaultAirflowRepository=apache/airflow --set images.airflow.repository=apache/airflow --set images.airflow.tag=master-python3.6-kubernetes -v 1 --set defaultAirflowTag=master-python3.6-kubernetes -v 1 --set config.api.auth_backend=airflow.api.auth.backend.default --dry-run
```

I get the following errors:
```
Error: values don't meet the specifications of the schema(s) in the following chart(s):
airflow:
- config.elasticsearch_configs.max_retries: Invalid type. Expected: string, given: integer
- config.elasticsearch_configs.timeout: Invalid type. Expected: string, given: integer
- config.scheduler.run_duration: Invalid type. Expected: string, given: integer
- config.scheduler.scheduler_heartbeat_sec: Invalid type. Expected: string, given: integer
- config.scheduler.statsd_port: Invalid type. Expected: string, given: integer
```

And indeed we have the following values:
```yaml
config:
  scheduler:
    scheduler_heartbeat_sec: 5
    statsd_port: 9125
    run_duration: 41460
  elasticsearch_configs:
    max_retries: 3
    timeout: 30
```

Should we allow primitive types as well (integer, number and boolean) ?

Comment (mik-laj) [reply chart/values.schema.json]: Yes. We should allow other types also.

**Comment by flvndh** [comment]
@mik-laj @dimberman Do you have any feedback on this PR ? Also, should I clean a bit the commit history ?

**Review by mik-laj — APPROVED** [review]

**Comment by mik-laj** [comment]
> Also, should I clean a bit the commit history ?

We use squash and merge so it's not important to us. If you wish, you can do so, but it is not required.

I can see your branch is a bit outdated. Can you do a rebase?

**Comment by mik-laj** [comment]
@dimberman Can you look at it?

**Comment by mik-laj** [comment]
@schnie Can I ask for review?

**Review by kaxil — APPROVED** [review]

**Pull request merged**